contracts/liquidity_hub/pool-network/incentive/src/claim.rs

```diff
@@ -0,0 +1,545 @@
+            continue;
+        }
+
+        flow.claimed_amount = flow.claimed_amount.checked_add(user_reward)?;
+
+        match &flow.flow_asset.info {
+            AssetInfo::NativeToken { denom } => messages.push(
+                BankMsg::Send {
+                    to_address: info.sender.clone().into_string(),
+                    amount: vec![Coin {
+                        amount: user_reward,
+                        denom: denom.to_owned(),
+                    }],
+                }
+                .into(),
+            ),
+            AssetInfo::Token { contract_addr } => messages.push(
+                WasmMsg::Execute {
+                    contract_addr: contract_addr.to_owned(),
+                    msg: to_binary(&cw20::Cw20ExecuteMsg::Transfer {
+                        recipient: info.sender.clone().into_string(),
+                        amount: user_reward,
+                    })?,
+                    funds: vec![],
+                }
+                .into(),
+            ),
+        }
+    }
+
+    // save the new flow state
+    //FLOWS.save(deps.storage, &flows)?;
+
+    LAST_CLAIMED_INDEX.save(deps.storage, info.sender.clone(), &env.block.time.seconds())?;
+
+    Ok(messages)
+}
+
+// TODO new stuff, remove/refactor old stuff
+
+/// Performs the claim function, returning all the [`CosmosMsg`]'s to run.
+pub fn claim2(
+    deps: &mut DepsMut,
+    env: &Env,
+    info: &MessageInfo,
+) -> Result<Vec<CosmosMsg>, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    let epoch_response: white_whale::fee_distributor::EpochResponse =
+        deps.querier.query_wasm_smart(
+            config.fee_distributor_address.into_string(),
+            &white_whale::fee_distributor::QueryMsg::CurrentEpoch {},
+        )?;
+
+    let current_epoch = epoch_response.epoch.id.u64();
+    let last_claimed_epoch = LAST_CLAIMED_EPOCH.may_load(deps.storage, &info.sender.clone())?;
+
+    if let Some(last_claimed_epoch) = last_claimed_epoch {
+        // if the last claimed epoch is the same as the current epoch, then there is nothing to claim
+        if current_epoch == last_claimed_epoch {
+            return Err(ContractError::NothingToClaim {});
+        }
+    }
+
+    // let user_share = get_user_share(&deps.as_ref(), info.sender.clone())?;
+
+    // calculate flow rewards
+    let mut messages: Vec<CosmosMsg> = vec![];
+    // let mut flows: Vec<Flow> = FLOWS
+    //     .may_load(deps.storage)?
+    //     .unwrap_or_default()
+    //     .into_iter()
+    //     //.filter(|flow| flow.start_timestamp <= env.block.time.seconds())
+    //     .filter(|flow| flow.start_epoch <= current_epoch)
+    //     .collect();
+    let mut flows: Vec<Flow> = FLOWS
+        .range(deps.storage, None, None, Order::Ascending)
+        .collect::<StdResult<Vec<((u64, u64), Flow)>>>()?
+        .into_iter()
+        .filter(|((start_epoch, _), _)| start_epoch <= &current_epoch)
+        .map(|((_, _), flow)| flow)
```

Comment:
This is sort of confusing but I get it, very slick !

Reply:
Idk how it could be simplified but I agree, not very pleasant to read

Reply:
Simplified it a bit, no more `(_,_)` but just `_`, take a look.